src/server.ts

```diff
@@ -749,3 +752,69 @@ function findCreatedAccountSequenceInTransactionMeta(
 
   throw new Error("No account created in transaction");
 }
+
+
+// TODO - remove once rpc updated to
+// append expiration entry per data LK requested onto server-side response 
+// https://github.com/stellar/soroban-tools/issues/1010
+function mergeResponseExpirationLedgers(ledgerEntriesResponse: SorobanRpc.RawGetLedgerEntriesResponse,
+  requestedKeys: xdr.LedgerKey[]
+): SorobanRpc.RawGetLedgerEntriesResponse {
+  const expirationKeyToRawEntryResult = new Map<String, SorobanRpc.RawLedgerEntryResult>();
+  const requestedKeyXdrs = new Set<String>(requestedKeys.map(requestedKey =>
+    requestedKey.toXDR('base64')));
+
+  (ledgerEntriesResponse.entries ?? []).forEach(rawEntryResult => {
+    if (!rawEntryResult.key) {
+      // don't interpret raw ledger entry data here, just pass it through to the outgoing response as-is
+      expirationKeyToRawEntryResult.set(Math.random().toString(), rawEntryResult)
+      return;
+    }
+    const parsedKey = xdr.LedgerKey.fromXDR(rawEntryResult.key, 'base64');
+    
+    if (parsedKey.switch().value !== xdr.LedgerEntryType.expiration().value ||
+        requestedKeyXdrs.has(rawEntryResult.key)) {
+      let keyHash = hash(parsedKey.toXDR()).toString()
+      let rawEntryResultLookup = expirationKeyToRawEntryResult.get(keyHash);
+
+      if (!rawEntryResultLookup) {
+        rawEntryResultLookup = rawEntryResult;
+        expirationKeyToRawEntryResult.set(keyHash, rawEntryResultLookup);
+      } 
+      rawEntryResultLookup.key = rawEntryResult.key;
+      rawEntryResultLookup.xdr = rawEntryResult.xdr;
+      rawEntryResultLookup.lastModifiedLedgerSeq = rawEntryResult.lastModifiedLedgerSeq;
+      return;
```

Comment:
Sorry, I'm having trouble following this logic. It seems like 

 * if the entry doesn't exist, you're setting `keyHash` to `rawEntryResultLookup` 
   - (which is the same as `rawEntryResult`))
 * then, if it _does_ exist, you are setting its fields to the same as those in `rawEntryResult` (i.e. overwriting what was there anyway)

Doesn't that mean that this whole sequence is always equivalent to just:

```suggestion
      expirationKeyToRawEntryResult.set(rawEntryResult);
      return;
```

Am I misunderstanding something here?

Reply:
if the order of GetLedgerEntriesResponse.entries was:
`[RawLedgerEntryResult.key=expirationKey, RawLedgerEntryResult.key=datakey]`

then the lookup gets populated with keyHash={expiration: xyz} on first iteration with the expiration entry on line 798, then the data entry comes through on second iteration, produces the same keyHash, gets that existing lookup value and on lines 784-786 its props are merged to form final object of {expiration, key, xdr, lastModifiedLedgerSeq}

i looked around a bit on merging or union of two objects with result of superset of keys, but didn't find much and settled on with this manual thing.

Reply:
ahhhh yes okay. so the spread operator (`...{}`) should help us do a merge, then: it will copy existing keys and you can add new ones

Reply:
yes, thanks for the [optimal suggeestion you made ](https://github.com/stellar/js-soroban-client/pull/161#discussion_r1355618090), which leveraged that and I re-wired to that.